torchdata/datapipes/iter/util/header.py

```diff
@@ -20,14 +21,26 @@ class HeaderIterDataPipe(IterDataPipe[T_co]):
     def __init__(self, source_datapipe: IterDataPipe[T_co], limit: int = 10) -> None:
         self.source_datapipe: IterDataPipe[T_co] = source_datapipe
         self.limit: int = limit
+        self.length = -1
 
     def __iter__(self) -> Iterator[T_co]:
         for i, value in enumerate(self.source_datapipe):
             if i < self.limit:
                 yield value
             else:
                 break
+        self.length = min(i + 1, self.limit)  # We know length with certainty when we reach here
```

Comment:
`i + 1` for the case when `source_datapipe` has a length smaller than the limit, so the loop exits with `i = len(source_datapipe) - 1`.

Reply:
There is a minor bug that `i` would be unavailable if `self.source_datapipe` is empty. Let's declare a variable before the iteration to represent `i`.

Reply:
Could we do it in a pythonic way
```suggestion
i = 0
for value in self.source_datapipe:
    i += 1
    if i < self.limit:
        yield value
    else:
        break
self.length = min(i, self.limit)
```